cheat-sheet.md

```diff
@@ -0,0 +1,250 @@
+3. This process repeats for each item of the list and returns the return value of the function once all items in the list have been iterated over.
+
+### foldLeft
+
+The first argument of foldLeft is a seed value `B` to be used for the first element, and the second argument is the function to apply.
+
+```
+def foldLeft[B](z: B)(op: (B, A) ⇒ B): B
+```
+
+Example:
+```
+List(1, 2, 3).foldLeft(20)(_ + _)
+// result = 6
+
+List(1, 2, 3).foldLeft(100)(_ + _)
+// result = 106
+```
+
+### foldRight
+The first argument of foldRight is a seed value `B` to be used for the first element, and the second argument is the function to apply.
+```
+def foldRight[B](z: B)(op: (A, B) ⇒ B): B
+```
+
+Example:
+```
+List(1, 2, 3).foldRight(0)(_ + _)
+// result = 6
+
+List(1, 2, 3).foldRight(100)(_ + _)
+// result = 106
+```
+
+### Comparison of fold, foldLeft and foldRight order:
+`foldLeft` starts on the left side (the first item) and iterates to the right; `foldRight` starts on the right side (the last item) and iterates to the left; `fold` has no particular order.
+
+```
+val list: List[Int] = List(1, 3, 5, 7, 9)
+
+list.foldLeft(0)(_ + _)
+// This is the only valid order
+0 + 1 = 1
+        1 + 3 = 4
+                4 + 5 = 9
+                        9 + 7 = 16
+                                16 + 9 = 25 // done
+
+list.foldRight(0)(_ + _)
+// This is the only valid order
+0 + 9 = 9
+        9 + 7 = 16
+                16 + 5 = 21
+                         21 + 3 = 24
+                                  24 + 1 = 25 // done
+
+list.fold(0)(_ + _) // 25
+// One of the many valid orders
+0 + 1 = 1    0 + 3 = 3             0 + 5 = 5
+        1            3 + 7 = 10            5 + 9 = 14    
+        1                    10          +         14 = 24
+        1                        +                      24 = 25 // done
+
+```
+
+### Algebraic Data Type (ADT)
+ADTs define a fixed set of all possible values of a given type. Two common classes of algebraic types are product types `AND` and sum types `OR`. 
+For sum types, the number of all its possible values is the sum or disjoint union of the number of all values of the two underlying types. 
+The values of a product type typically contain several values, called fields. For products types, the number of its possible values is the product of the fields.
+
+Values of algebraic types are analysed with pattern matching, which identifies a value by its constructor or field names and extracts the data it contains.
+
+### Sealed trait
+When defining an algebraic data type using sealed traits, it allows the compiler to exhaustively check the possible cases in match expressions. The compiler will emit a warning if the match expression isn’t exhaustive. The compiler knows all of the subtypes of the trait that can possibly exist as they can only be extended in the file. 
+
+```
+sealed trait Boolean
```

Comment:
Perhaps use `Colour` with 2 or 3 colours as an example. 

`True` and `False` don't exist in Scala, they're actually `true` and `false` - so this might confuse.

Reply:
Umm I would disagree with this on @wjlow , I like the Boolean example as it is based on a well know case (only two cases valid). What I would suggest is renaming the trait to something like `BooleanValue` or `MyBoolean` to make clearer that it is just an example.

Reply:
Good point! `BooleanType` as the `sealed trait` would be sufficient :)